### PR TITLE
Make /listLanguageNames non-blocking

### LINKED ISSUE #34: langnames calls should be non-blocking
if getLocalizedLanguages reads from a damaged disk/db, it can block other requests. It should use the same non-blocking yield pattern as translation does.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #34

### CODE REVIEW

**Comment by Androbin** [comment]
Hope I got the syntax right.

**Comment by Androbin** [comment]
@unhammer Does this match your intent?

**Comment by unhammer** [comment]
Looks right; have you tested it? 

Actually, I'm not sure how to do that easily; #34 was prompted by langnames.db being on an NFS drive, and sometimes access to that would be super-slow; maybe it's possible to create a very slow loopback filesystem?

**Comment by Androbin** [comment]
We could try some of the suggestions here:
https://askubuntu.com/questions/13590/can-i-simulate-a-slow-hard-drive

I will try out some of them.

**Comment by Androbin** [comment]
@unhammer Where can I get `langNames.db` from?

**Comment by Androbin** [comment]
I figured that `time.sleep` in the right places should do the job.

But I don't get quite the behavior I expected:
* Using one core, both `analyze` and `listLanguageNames` are blocking
* Using four cores, neither `analyze` nor `listLanguageNames` are blocking

I am testing current `master` btw.

**Comment by unhammer** [comment]
They should be non-blocking on one core (and there's really no need to run apy with `-j>1`, I think we might want to deprecate the `-j` option at some point).

I think `time.sleep` (which happens in the Python process) should force them to be blocking, which isn't quite what we want :) We have to make the stuff outside of the Python process block.

For `analyze`, you can test it by sending through large texts and lots of small at the same time, see https://asciinema.org/a/155212 Here the large text is processing concurrently with the small ones. There's only one Python process running, but that process can wait for several backgrounded `apertium` processes.

`analyze`  is simple to test for concurrency since you just send small and large input. But `listLanguageNames` is (under normal circumstances) so fast that it's hard to check that it's not blocking other requests. So we have to force abnormal circumstances to test it.

**Comment by Androbin** [comment]
@unhammer Could you please re-iterate why we can't simulate a slow `getLocalizedLanguages` by injecting a `time.sleep` ?

**Comment by Androbin** [comment]
@unhammer Could you please re-iterate why we can't simulate a slow `getLocalizedLanguages` by injecting a `time.sleep` ? Could we use a large `for x in range(999999999)` loop instead?

**Comment by unhammer** [comment]
> @unhammer Could you please re-iterate why we can't simulate a slow `getLocalizedLanguages` by injecting a `time.sleep` ? Could we use a large `for x in range(999999999)` loop instead?

No, Python code will still happen in the same Python thread, it would test something different. An async coroutine can only give back control to the loop at the places where you've put `yield` in front of a call. If you run a long for loop, it should block for the duration of that loop (that's to be expected), while if you put a `yield` inside the body of the for-loop, it should not block. But that wouldn't test what happens when the database connection hangs.

Looking more closely at `getLocalizedLanguages`, I see three possible lines that could hang:

```
            langNameDBConn = sqlite3.connect(dbPath)
            c = langNameDBConn.cursor()
[…]
    languageResults = c.execute('SELECT * FROM languageNames WHERE lg=?', (locale, )).fetchall()
```

I'll bet the first one is where I had my NFS issue. Since it's using the sqlite3 library and not a shell command, I'm now starting to doubt it's easy to yield correctly there. But in any case, we have to fake a slow file connection in order to test whether it's working or not.

**Comment by Androbin** [comment]
Okay, I've moved the database to my Google Drive, mounted it via a super-duper-slow client and created a link from the original path to the new one. If this isn't slow enough, I can just unplug my router.

**Comment by Androbin** [comment]
In case we need to do that again: It took over 16 seconds to load. Yeah!

**Comment by unhammer** [comment]
That seems like a great test then :) If we're able to do a short `/analyze` during those seconds (with `-j1`), it's non-blocking. But I'm guessing it's not possible to just put a "yield" around those sqlite calls … @sushain97 suggestions?

**Comment by Androbin** [comment]
> But I'm guessing it's not possible to just put a "yield" around those sqlite calls

http://www.tornadoweb.org/en/stable/guide/coroutines.html#calling-blocking-functions

**Comment by Androbin** [comment]
> `sqlite3.ProgrammingError`: SQLite objects created in a thread can only be used in that same thread.

We have to put all SQL calls in **one** block, wrapped with `ThreadPoolExecutor`.

**Comment by Androbin** [comment]
Heureka! It works!

File: util.py [hunk 1]
```diff
@@ -39,39 +43,55 @@ def toAlpha3Code(code):
...
+
+    return cursor.execute('SELECT * FROM languageNames WHERE lg=?', (locale, )).fetchall()
+
+
+@gen.coroutine
+def getLocalizedLanguages(locale, dbPath, languages=[]):
+    languageResults = yield ThreadPoolExecutor(1).submit(getLanguageNames, locale, dbPath)
+
+    if not languageResults:
+        logging.error('Failed to locate language name DB: %s' % dbPath)
```
Comment (sushain97) [inline util.py]: Do we really need this error twice?

File: util.py [hunk 2]
```diff
@@ -39,39 +43,55 @@ def toAlpha3Code(code):
...
         if os.path.exists(dbPath):
             langNameDBConn = sqlite3.connect(dbPath)
-            c = langNameDBConn.cursor()
+            cursor = langNameDBConn.cursor()
         else:
             logging.error('Failed to locate language name DB: %s' % dbPath)
-            return {}
+            return None
+
+    return cursor.execute('SELECT * FROM languageNames WHERE lg=?', (locale, )).fetchall()
```
Comment (sushain97) [inline util.py]: Could this ever result in multiple concurrent threads accessing the same cursor? The cursor isn't thread-safe.

Comment (unhammer) [reply util.py]: I'm guessing also multiple processes (starting servlet with -j2) would mess this up.

File: util.py (hunk 1 above)
Comment (Androbin) [reply util.py]: Sorry, forgot to clean up the artifacts.

File: util.py (hunk 2 above)
Comment (Androbin) [reply util.py]: Okay, then let's make `ThreadPoolExecutor` a global and keep it single-threaded, so that we can guarantee only one access at a time.

**Comment by sushain97** [comment]
It looks good to me now. @unhammer any thoughts?

**Comment by Androbin** [comment]
This PR also resolves the problem of the cursor being undefined on subsequent runs.

**Comment by unhammer** [comment]
LGTM :)

**Pull request merged**